### DISCUSSION TITLE
Optimisation for tracking small features on small organisms

### DISCUSSION THREAD

**Discussion by bClemot-Sc** [talmolab/sleap#2735] [Help!]
Hello SLEAP community, 

I am opening a discussion as I am currently using SLEAP to track **white markers on single freely behaving dung beetles**. The main challenge I have been experiencing is that my organism is fairly **small** in the field of view, and so are the features I am tracking on it. See here an example of a frame:

<img width="1920" height="1080" alt="vlcsnap-2026-06-03-16h25m12s063" src="https://github.com/user-attachments/assets/e7874a78-8afd-4385-a2f9-041f4bad904d" />

Even if the features are small, my videos are in **1920*1080** resolution and filmed at **50 fps**. The markers end up being approximately **5 pixels wide** and **spaced by 2 to 5 times this width**. Therefore, I managed to easily **label a total of 400 frames** at the moment.

My main problem now is in tweaking the parameters in order to get the best results from SLEAP. After several bad results with a single animal model, I tried using a **top-down model** constrained to one instance, so that the centroid model would locate the beetle, crop around it and then feed this for the second model to learn the marker's position. From this, I got half-good results, with the predictions still being very imprecise in some instances:

- **Here is for the centroid model:** In which I always have this weird behaviour of the validation going down then up before stabilising (I tried increasing the amount of validation data, and nothing changed).

[training_config_centroid1.txt](https://github.com/user-attachments/files/28556326/training_config_centroid1.txt)

<img width="869" height="676" alt="image" src="https://github.com/user-attachments/assets/12bf284c-b102-434e-b9f7-9dd007b78d63" />

//
- **Here is for the centered instance model:** In which I have very low learning that leads to overfitting.

[training_config_centered_instance1.txt](https://github.com/user-attachments/files/28556442/training_config_centered_instance1.txt)

<img width="1106" height="676" alt="image" src="https://github.com/user-attachments/assets/4223682d-b71e-4e30-b2a2-b26fc41f2630" />

From this model, I got results where the model would easily get confused between dots and also would position unprecisely.

I tried several tweaking, such as increasing the receptive field through the **Max Stride** but this didn't helped much. I tried to keep the **scaling at 1** to not lose resolution on my small markers. Notably, I also tried to reduce the **Output Stride** to 1 to not lose reslution relative to my small markers size. But in this scenario I obtain way worse results:

- **Here for the centered instance model:** Where I got very bad results with no learning.

[training_config_centered_instance.txt](https://github.com/user-attachments/files/28556717/training_config_centered_instance.txt)

<img width="1128" height="580" alt="image" src="https://github.com/user-attachments/assets/2f920689-fc92-4395-9695-54b5552a21ae" />

I have also tried with an **Output Stride of 2**, and got in-between results. I am planing to try more and more with different setup, probably with different **Max Stride** and **Sigma**. But I'd like to ask for any expert knowledge and advice to improve my model learning in my situation.

Thank you very much and very happy to provide more details, and send the results of any suggestion you'd make.

**Comment by talmo** [comment]
Hi @bClemot-Sc,

Thanks for the thorough recap of what you've tried so far! I think you're doing everything right on the model config side, but the validation curves indeed look strange.

Would you mind sharing your labels so we can have a look to help diagnose? You can upload the package file (Predict -> Export labels package) to https://slp.sh and just let us know and we'll take a peek :)

Cheers,

Talmo

**Comment by bClemot-Sc** [comment]
Hi @talmo,

Thank you very much for having a look at my case!

Following your asking, I have exported the labelled + suggested frames here: [https://slp.sh/awXLFn](url).
Happy to provide additional details or tests!

Best,

**Comment by talmo** [comment]
Hi @bClemot-Sc,

Thanks for sending the labels — that let us actually **reproduce your setup and test fixes on
your own data** rather than just guess. We trained a handful of centered-instance models on a
365/41 split of your frames (single TITAN RTX, 50 epochs each) and measured keypoint accuracy.
Good news: this is all fixable on the config side, and we found exactly what's hurting you. 🎉

## First, an important measurement note

You've been comparing models by their **validation loss curves** — but loss is *not comparable*
across different `sigma`/`output stride` settings, because the target confidence maps literally
have different scales. That's a big part of why your comparisons looked confusing. The honest
metric is **keypoint error in pixels** on held-out frames, which is what we used below.

## What we changed, and what each change did

Starting from your best config and changing **one thing at a time** (all on your data, same
held-out val set). The number that matters is **p90 error** — the *worst* markers, i.e. your
"confused dots." The median barely moves because most markers are already easy:

| change | crop | out-stride | σ | max-stride | p90 error | % within 5px |
|---|---|---|---|---|---|---|
| **your baseline** | 112 | 4 | 1.5 | 8 | 12.8 px | 83% |
| + crop size 160 | **160** | 4 | 1.5 | 8 | 8.1 px | 85% |
| + out-stride 2, σ 3 | 160 | **2** | **3** | 8 | 7.8 px | 87% |
| + max-stride 16 | 160 | 2 | 3 | **16** | **3.7 px** | **93%** |

**End result: worst-case error dropped from 12.8 px → 3.7 px (−71%), and markers within 5 px
went from 83% → 93%.** On the hardest frames the difference is dramatic — see the attached
figure: your baseline (left) swaps/misplaces the outer dots by 10–90 px, the tuned model
(right) lands on them.

<img width="1143" height="1787" alt="preds_A_vs_D" src="https://github.com/user-attachments/assets/7adb25f4-4346-4bd0-9aad-3a93e5b08687" />

The three levers, in order of impact:

<img width="1788" height="705" alt="experiment_results" src="https://github.com/user-attachments/assets/02ac636a-a882-4558-969b-292710ee4065" />


### 1. Max stride 8 → 16 (biggest single win: p90 7.8 → 3.7 px)
Your markers are visually identical, so the only way the network can tell them apart is the
**overall geometry of the constellation** — which means it has to "see" the whole beetle at
once. Bumping max stride to 16 widens the receptive field to cover the ~100 px constellation,
and it cut the worst-case error roughly in half. (This is why your earlier max-stride
experiments "didn't help much" on their own — the crop and stride issues below were masking it.)
We also tried max stride **32** — no further gain over 16, so 16 is the sweet spot here.
(For reference, **16 is also the default SLEAP suggests for the centered-instance model** — it
looks like yours had been set to 8.)

### 2. Crop size 112 → 160 (p90 12.8 → 8.1 px, −36%)
This is the one we'd never have caught without your file. We measured every labeled beetle: the
constellation reaches up to **67 px from its center**, but your crop of 112 only extends 56 px
from center. So on **~1 in 3 frames** (worse because you use full ±180° rotation) the outer
markers fall *outside the crop* and the model literally can't see them during training. Crop
160 covers 100% of your beetles at any rotation — and it's exactly what SLEAP's own config
recommender picks for your data (~1.5× the largest animal). This fix mostly rescues the outer
markers (Head_F*, Abdomen_B*), which is where your worst errors were.

### 3. Output stride 4 → 2 (the precision knob)
Worth clearing up how `sigma` works, since it explains your two failed experiments. In
SLEAP-NN, **sigma is defined at the output-grid scale** — the target blob is always ~`sigma`
grid-cells wide, and its *physical* size in the image is `sigma × output_stride`. So your
OS4/σ1.5 and a OS2/σ3 actually make the *same* ~6 px target; what differs is the **output grid
resolution**, which sets how precisely peaks can be pinned down.
- **Output stride 4**: localizes on a 4×-coarser grid → less precise (you can see it in the
  table: fewer markers land within a tight tolerance).
- **Output stride 1** (your "no learning"): at full resolution a σ1.5 blob is tiny in a huge,
  almost-entirely-empty output, so the loss is minimized by just predicting near-zero
  everywhere. (You *could* rescue OS1 with a much larger sigma, but it's wasteful — OS2 is the
  clean fix.)
- **Output stride 2 (σ ≈ 2.5–3)** is the sweet spot: fine enough to localize precisely, dense
  enough to train cleanly.

### 4. Intensity augmentation — turn it on (you currently have it off)
In our same-videos test this didn't change the numbers, **because its benefit is generalizing
to *new* lighting**, which a held-out split of the same videos can't show. Your footage is
outdoor with strong, moving sun/shadow, so for applying the model across your full recordings
we'd still strongly recommend enabling brightness / contrast (and a little noise). No downside,
real upside on unlabelled video.

## Recommended config (centered-instance)

| setting | value |
|---|---|
| Input scaling | 1.0 |
| **Crop size** | **160** |
| **Output stride** | **2** |
| **Sigma** | **3.0** (sweep 2.5–3.5; this is in output-grid units) |
| **Max stride** | **16** |
| Filters | 24–32 |
| Rotation | ±180° (keep) |
| Brightness / Contrast aug | ON (~0.5) |
| Batch / LR | 4 / 1e-4 (fine as-is) |

Your **centroid model is healthy** — the "down then up then stabilize" wiggle is a normal
early-training transient, and it localizes the beetle reliably. No changes needed there (you
can add the same intensity augmentation).

If a few outer markers still occasionally swap after this, the next lever is a handful more
labeled frames emphasizing unusual orientations / partial occlusions — but the config changes
above should get you most of the way. Let us know how it goes, and feel free to share the new
curves!

Cheers,

:heart: Talmo & Claude :robot:

<details>
<summary><b>Extended technical analysis (measurements + full ablation)</b></summary>

Measured across 395 user instances in your package (full-res px): constellation bbox up to
104×107, **centroid→farthest-marker radius median 50 / p90 61 / max 67**; nearest-neighbor
marker spacing median 17 px (≈15% < 10 px). So with crop 112 (half-extent 56 px) **33.4% of
instances have ≥1 marker beyond the crop** at some rotation; crop 128 covers 98.5%, crop 160
covers 100%. (Mechanism: the centered-instance pipeline crops at `crop_size×√2` for rotation
headroom, rotates, then re-crops to `crop_size` centered on the centroid — so the √2 only
prevents black corners; the effective field of view stays `crop_size`.)

Full ablation (50 epochs, identical 365/41 split, scale 1.0, batch 4, lr 1e-4, ±180° rotation;
keypoint error in px on the held-out val set):

| run | crop | OS | σ | max-stride | aug | median | mean | p90 | %≤5px | %≤10px |
|---|---|---|---|---|---|---|---|---|---|---|
| A baseline | 112 | 4 | 1.5 | 8 | off | 1.75 | 5.10 | 12.78 | 82.5 | 88.8 |
| B +crop160 | 160 | 4 | 1.5 | 8 | off | 1.68 | 4.19 | 8.13 | 85.3 | 91.8 |
| C +OS2/σ3 | 160 | 2 | 3 | 8 | off | 1.63 | 4.56 | 7.84 | 87.3 | 92.0 |
| E +maxstride16 | 160 | 2 | 3 | 16 | off | 1.14 | 2.48 | 3.65 | 93.0 | 96.0 |
| D +intensity aug | 160 | 2 | 3 | 16 | on | 1.09 | 2.03 | 3.75 | 93.0 | 97.8 |

Confmap-scale note: SLEAP-NN renders targets with `make_confmaps(points, xv, yv, sigma *
output_stride)` on a grid sampled every `output_stride` input px — so the gaussian is `sigma`
grid-cells wide (constant) and `sigma × output_stride` **input px** wide. Implications: target
peaks are always well-resolved in the grid regardless of OS; OS sets **localization
granularity** (coarser grid at OS4 → worse fine precision, visible as lower %≤3px for the OS4
runs A/F). The OS1 "no learning" is a sparsity failure: at full resolution a small σ blob is a
negligible fraction of the output → MSE collapses to the all-zero solution. Physical target
sizes here: A/B/C/D/E all ≈ 6 px (σ×OS = 1.5×4 or 3×2); F ≈ 10 px (2.5×4), broader → less
precise, consistent with F's worst %≤3px (59%).

Caveat on understatements: val frames are not rotated, so crop clipping at eval is only ~2.2%
vs ~33% during training — the real inference benefit of crop 160 is larger than the val p90
shows. Augmentation's benefit (new-lighting generalization) is likewise not captured by a
same-videos split.

</details>

**Comment by bClemot-Sc** [comment]
Hi @talmo!

**Thank you so much for your wonderful investigation and help! I really appreciate!**

Following your suggestions, I was able to replicate your observations and obtain similar results on my centred-instance model training! Which is really positive! Tweaking the parameters as you advised and adding Brightness and Contrast data amplification seems to have helped a lot.

I then ran some inference on my videos, and for some parts, got very good results. However, I obtained a lot of confusion and jittering between the markers, and even some complete sections where the model is unable to detect the markers at all.

<img width="1297" height="908" alt="Screenshot from 2026-06-07 14-33-19" src="https://github.com/user-attachments/assets/82e79ae0-091e-47e8-ba7b-1c1cd56e70a8" />

I also tried running inferences with the Flow Tracking Method and an Elapsed Frames Window of 10. While it might have resolved some jittering, I still got a significant portion of my videos with the model unable to detect the markers.

<img width="1459" height="1094" alt="Screenshot from 2026-06-07 13-52-12" src="https://github.com/user-attachments/assets/e1d2e2a4-8223-480e-8a8a-84be907f5b2e" />

I am planning to first check if the Centroid model might sometimes get confused between the dung ball and the actual animal, explaining why some instances have almost no markers detected. Then I will try to label more frames as you suggested!


I'm happy to get any other suggestions you might think about, and I will keep this chat updated with my progress! 

**Thank you very much!!!** :)
